cli_tools/gce_vm_image_import/importer/process_planner.go

```diff
@@ -0,0 +1,123 @@
+
+package importer
+
+import (
+	"errors"
+	"log"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/disk"
+	daisy_utils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisy"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	"github.com/GoogleCloudPlatform/compute-image-tools/proto/go/pb"
+	"google.golang.org/api/compute/v1"
+)
+
+// processPlanner determines which actions need to be completed to during processing
+// to ensure a disk is bootable on GCE.
+type processPlanner interface {
+	plan(pd persistentDisk) (*processingPlan, error)
+}
+
+// newProcessPlanner returns a processPlanner that uses disk.Inspector to backfill
+// information that is missing from the arguments (such as: which  OS is on the disk, or whether
+// UEFI is required to boot the disk).
+func newProcessPlanner(args ImportArguments, diskInspector disk.Inspector) processPlanner {
+	return &defaultPlanner{args, diskInspector}
+}
+
+// processingPlan describes the metadata and translation steps that need to be performed
+// against a disk to make it bootable on GCE.
+type processingPlan struct {
+	requiredLicenses        []string
+	requiredFeatures        []*compute.GuestOsFeature
+	translationWorkflowPath string
+}
+
+// metadataChangesRequired returns whether metadata needs to be updated on the
+// GCE disk resource object.
+func (plan *processingPlan) metadataChangesRequired() bool {
+	return len(plan.requiredLicenses) > 0 || len(plan.requiredFeatures) > 0
+}
+
+type defaultPlanner struct {
+	args          ImportArguments
+	diskInspector disk.Inspector
+}
+
+func (p *defaultPlanner) plan(pd persistentDisk) (*processingPlan, error) {
+	// Don't run inspection if the user specified a custom workflow.
+	if p.args.CustomWorkflow != "" {
+		return &processingPlan{translationWorkflowPath: p.args.CustomWorkflow}, nil
+	}
+
+	inspectionResults, inspectionError := p.inspectDisk(pd.uri)
+
+	osID := p.args.OS
+	requiresUEFI := p.args.UefiCompatible
+	if inspectionError == nil && inspectionResults != nil {
+		if osID == "" && inspectionResults.OsCount == 1 && inspectionResults.GetOsRelease() != nil {
+			osID = inspectionResults.GetOsRelease().CliFormatted
+			if p.args.BYOL {
+				osID += "-byol"
+			}
+		}
+
+		if !requiresUEFI {
+			hybridGPTBootable := inspectionResults.GetUefiBootable() && inspectionResults.GetBiosBootable()
+			if !p.args.UefiCompatible && hybridGPTBootable {
+				log.Printf("This disk can boot with either BIOS or a UEFI bootloader. The default setting for booting is BIOS. " +
+					"If you want to boot using UEFI, please see https://cloud.google.com/compute/docs/import/importing-virtual-disks#importing_a_virtual_disk_with_uefi_bootloader'.")
+			}
+			requiresUEFI = inspectionResults.UefiBootable && !hybridGPTBootable
+		}
+	}
+
+	if osID == "" {
+		return nil, errors.New("Could not detect operating system. Please re-import with the operating system specified")
```

Comment:
A link to docs would be useful here since the error is terminal

Reply:
I like the overall idea. What docs were you thinking?

A potential issue is that the code will have a few different consumers:

1. Client: gcloud, web UI, direct cloud build invocation
2. Tool: OVF import, image import

Reply:
Differentiating between clients is already covered in the docs. E.g. for image import, this link should work: https://cloud.google.com/compute/docs/import/importing-virtual-disks#bootable

Differentiating between tools is an issue. Error message should be customized, either by letting calling code to wrap it, or customize processors/planner to have tool specific error messages.

Reply:
Added the image import link for now; you have a good point above customizing error messages, so I'll keep that in mind when incorporating this into OVF import!